[Windows] Add Antrea Windows design doc (#751)
vmware-tanzu/antrea: docs/windows-design.md

# Running Antrea on Windows

> Antrea supports running on Windows worker Node. In this case, the Windows Node should

Reviewer: on Windows worker Nodes

> Antrea supports running on Windows worker Node. In this case, the Windows Node should

Reviewer: s/should/must

> be a worker in the cluster, while the master Node must be Linux. On Windows Node, 

Reviewer: On Windows Nodes,

Antrea sets up an overlay network to forward packets between Nodes and implements
NetworkPolicies. 

## Design

On Windows, the Host Networking Service (HNS) is a necessary component to support container

> networking. For Antrea on Windows, “Transparent” mode is chosen for the HNS network. In this

Reviewer: you may want to fix your quotes, it seems you copy-pasted them from somewhere

mode, containers will be directly connected to the physical network through an **external** Hyper-V switch. 

> OVS is working as the extension of the external Hyper-V switch, which is bound with the HNS 

Reviewer: is "bound" the best word here? I thought the Hyper-V switch was *created* by HNS

I would suggest the following but let me know if I misunderstood: "OVS is working as a forwarding extension for the external Hyper-V switch which was created by HNS"

> Network. Hence, the packets that are sent from/to the containers could be processed by OVS.

Reviewer: s/could/can

> The network adapter used in the HNS Network is also added into the OVS bridge as the uplink

Reviewer: s/added into/added to

> interface. An interface for the OVS bridge is created, and the original networking

Reviewer: s/An interface/An internal interface

...unless I am confused

> configuration(e.g., IP, MAC and route entries) on the host network adapter are moved to

Reviewer: ```suggestion
configuration (e.g., IP, MAC and routing entries) on the host network adapter is moved to
```

this new interface. Some extra OpenFlow entries are needed to ensure the host traffic can be
forwarded correctly.

<img src="/docs/assets/hns_integration.svg" width="600" alt="HNS Integration">

> SNAT based on OpenFlow is needed to make sure the containers could access the external address.

Reviewer: s/could/can

> The SNATed address is using the IP configured on the OVS bridge. Some assisted OpenFlow entries are

Reviewer: not sure what you meant to say with "assisted" here

maybe: "Some additional OpenFlow entries are installed to assist in identifying and forwarding the external traffic."

installed to help identify and forward the external traffic.

> kube-proxy is running on each Windows Node to provide Kubernetes Service. The Service traffic is

Reviewer: s/to provide Kubernetes Service/to implement Kubernetes Services ?

> kube-proxy is running on each Windows Node to provide Kubernetes Service. The Service traffic is

Reviewer: s/The Service traffic/Service traffic

> output to the host from gw0, and captured by kube-proxy. kube-proxy sets up a new connection to

Reviewer: s/from/on?

> the backend endpoint, and forwarded the request using this new connection. The forwarded request

Reviewer: s/and forwarded the request/and forwards the request

> enters OVS pipeline again and output to the actual Pod. To be compatible with OVS, kube-proxy on

Reviewer: "enters the OVS pipeline again and is output to the actual Pod"

Is it actually output to the Pod? Or it it forwarded to the Pod? If the Pod is remote, I don't think "output to the Pod" is accurate

> Windows must be configured to run in the **userspace** mode, and a specific network adapter is 

Reviewer: s/in the **userspace** mode/in **userspace** mode

> required to be provided on which the Service IP address is configured.

Reviewer: ```suggestion
required, on which Service IP addresses will be configured by kube-proxy.
```

### HNS Network configuration

> HNS Network is created at Antrea Agent initialization phase, and it should be created before creating 

Reviewer: s/is created at Antrea Agent initialization phase/is created during the Antrea Agent initialization phase

> OVS bridge. This is because OVS is working as the Hyper-V Switch Extension, and the ovs-vswitchd process

Reviewer: s/before creating OVS bridge/before the OVS bridge

> OVS bridge. This is because OVS is working as the Hyper-V Switch Extension, and the ovs-vswitchd process

Reviewer: > the ovs-vswitchd process cannot run until the Hyper-V Switch is created

This is confusing to me because it seems `PrepareNode.ps1` takes care of starting ovs-vswitchd (right after installing OVS) and that happens before the HNS network is created?

https://github.com/vmware-tanzu/antrea/blob/master/hack/windows/Install-OVS.ps1#L161

> cannot run until the Hyper-V Switch is create and the OVS Extension is enabled on this Switch. 

Reviewer: s/is create/is created

When creating the HNS Network, the local subnet CIDR and the uplink network adapter are required.
Antrea Agent finds the network adapter from the Windows host using the Node internal IP as a filter,

> and retrieves the local Subnet CIDR from Node spec. 

Reviewer: s/from Node spec/from the Node spec

> After the HNS Network is created, OVS extension should be enabled at once.

Reviewer: s/OVS extension should be enabled at once./the OVS extension should be enabled at once on the Hyper-V switch/

### Container network configuration

[**host-local**](https://github.com/containernetworking/plugins/tree/master/plugins/ipam/host-local)
plugin is used to provide IPAM for containers, and the address is allocated from the subnet CIDR
configured on the HNS Network. 

Windows HNS Endpoint is leveraged as the vNIC for each container. A single HNSEndpoint with the

> IP allocated from IPAM plugin is created for each Pod. The HNSEndpoint should bw attached on all
> containers in the same Pod to ensure thw network configuration could be correctly accessed(this

Reviewer: ```suggestion
IP allocated by the IPAM plugin is created for each Pod. The HNS Endpoint should be attached to all
containers in the same Pod to ensure that the network configuration can be correctly accessed (this
```

> operation is to make sure DNS configuration is readable from all containers).

Reviewer: s/DNS configuration/the DNS configuration

 
One OVS port with the same name as the HNS Endpoint is also needed, which is helping to control

> the container traffic with OpenFlow rules. OpenFlow entries are installed to implement connectivity

Reviewer: s/which is helping to control the container traffic with OpenFlow rules/in order to handle container traffic with OpenFlow rules

> between Pod and Pod, Pod and external. The service traffic is handled by kube-proxy, and a specific

Reviewer: suggestion: to implement Pod-to-Pod and Pod-to-external connectivity

> between Pod and Pod, Pod and external. The service traffic is handled by kube-proxy, and a specific

Reviewer: Service traffic is handled by ...

Openflow rule is installed to output these packets to the gateway port.

> CNIAdd request might be called multiple times for a single Pod, this is because kubelet on Windows

Reviewer: ```suggestion
The CNIAdd request might be called multiple times for a given Pod. This is because kubelet on Windows
```

assumes CNIAdd is an idempotent event, and it uses this event to query the Pod networking status.

> Antrea needs to identify the container type(sandbox or workload) from CNIAdd request: 1) creates
> the HNSEndpoint only when the request is for the sandbox container, 2) attaches the HNSEndpoint no
> matter it is sandbox container or workload container.

Reviewer: ```suggestion
Antrea needs to identify the container type (sandbox or workload) from the CNIAdd request:
 * we create the HNS Endpoint only when the request is for the sandbox container
 * we attach the HNS Endpoint no matter whether it is a sandbox container or a workload container.
```

### Gateway port configuration

> Gateway port is created in Antrea Agent initialzation phase, and the address of the port should be
> the first IP in the subnet. The port is an OVS internal port and default name is "gw0".

Reviewer: ```suggestion
The gateway port is created during the Antrea Agent initialization phase, and the address of the interface should be
the first IP in the subnet. The port is an OVS internal port and its default name is "gw0".
```

Openflow rule is installed to output these packets to the gateway port.

CNIAdd request might be called multiple times for a single Pod, this is because kubelet on Windows
assumes CNIAdd is an idempotent event, and it uses this event to query the Pod networking status.
Antrea needs to identify the container type(sandbox or workload) from CNIAdd request: 1) creates
the HNSEndpoint only when the request is for the sandbox container, 2) attaches the HNSEndpoint no
matter it is sandbox container or workload container.

### Gateway port configuration

Gateway port is created in Antrea Agent initialzation phase, and the address of the port should be
the first IP in the subnet. The port is an OVS internal port and default name is "gw0".

Gateway port is used to help implement L3 connectivity for the containers, including Pod to external,
Pod to Service, and node to Pod. For Pod to external, and Pod to Service cases, openFlow entries are
installed to output these packets to the host from the gateway port. To ensure the packet is forwarding
correctly on the host, the IP-Forwarding feature should be enabled on the network adapter of the

> gateway port.

Reviewer: The gateway port is used to help implement L3 connectivity for the containers, including Pod-to-external,
Pod-to-Service, and Node-to-Pod. For the Pod-to-external and Pod-to-Service cases, OpenFlow entries are
installed in order to output these packets to the host on the gateway port. To ensure the packet is forwarded
correctly on the host, the IP-Forwarding feature should be enabled on the network adapter of the
gateway port.

and the address is the one configured on the OVS bridge. 

### OVS bridge interface configuration

Since OVS is also responsible to take charge of the network of the host, an interface for the OVS bridge
is required on which the host network settings are configured. It is created and enabled when creating
the OVS bridge, and the MAC address should be changed to be same as the uplink interface. Then the IP 
address and the route entries originally configured on the uplink interface should also be migrated to
the interface.

The packet that is sent to/from the Windows host should be forwarded on this interface. So the OVS bridge
is also thought as one source from where the packet enters the OVS pipeline pipeline. A special ofport
number 65534(named as LOCAL) for the OVS bridge is used in OpenFlow spec. 

In OVS `Classifier` table, new OpenFlow entries are needed to match the packets from this interface.
There are two kinds of packets entering OVS pipeline from this interface: 1) from Windows host to external
address, 2) from Windows host to local Pods(e.g., this happens if the traffic is initiated from a local

> Pod to a NodePort Service).

Author: This is a corner case for traffic that Pod is accessing NodePort Service, and the chosen Node is where the client Pod is locating on. As the traffic is using Node's internal IP as the destination, SNAT is not taken on the packet. We observe kube-proxy on the Windows host receives the request and forwards the packet to OVS back from the bridge interface. Hence, we have this OpenFlow entry (https://github.com/vmware-tanzu/antrea/blob/master/pkg/agent/openflow/pipeline.go#L724) to ensure the packet can be forwarded to the Pod but not output to the uplink interface.

following changes on OpenFlow entries are needed:

* Pick out the packet that is sent from local Pod to external address and set the SNAT mark in the register
 in L3Forwarding table. The packet should have these characters: 1) it enters the OVS pipeline from local Pod,
 2) the destination IP address is not in any of these conditions: cluster IP(local or any of the remote Subnet
  CIDR), or Node internal IP, 3) the traffic is not for Service.
* Implement SNAT rules and commit the new connection to the ct context in ConntrackCommit table. The SNAT
 rule should change the source address to the Node internal IP.
* Resubmit the IP packet to the Conntrack table if it is from the uplink interface in Classifier table. This
 is to ensure the reply packets from external address to the Pod could be forwarded correctly.
* Ensure the packet could be handled with NAT configurtaion when entering the ct zone in Conntrack table.
* Rewrite dMAC with the global virtual MAC(aa:bb:cc:dd:ee:ff) in ConntrackState table if it is from the uplink
 interface and having the SNAT ct_mark.
* Resubmit the packets to DNAT table in ConntrackState table if they have the SNAT ct_mark but not from the

>  uplink interface. 

Reviewer: ```suggestion
* Rewrite the dMAC with the global virtual MAC (aa:bb:cc:dd:ee:ff) in the `ConntrackState` table if the packet is from 
   the uplink interface and has the SNAT ct_mark.
* Resubmit the packets to the `DNAT` table in the `ConntrackState` table if the packet has the SNAT ct_mark but 
   is not from the uplink interface. 
```

table=31, priority=200, ct_state=-new+trk,ct_mark=0x40,ip actions=resubmit(,40)
table=31, priority=200, in_port=$uplink,ip actions=output:br-int

Table70:
// Resubmit the packet to L2ForwardingCalculation table if it is for Service traffic. 
table=70, priority=200, ip,reg0=0x2/0xffff,ct_mark=0x20 actions=resubmit(,80)
// Resubmit the packet to L2ForwardingCalculation table if it is traffic from Pod to the Node.
table=70,priority=190,ip,reg0=0x2/0xffff,nw_dst=$local_nodeIP,actions=resubmit(,80)
// Resubmit the packet to L2ForwardingCalculation table if it is traffic to the local Pods
table=70,priority=190,ip,nw_dst=10.10.0.0/24,actions=resubmit(,80)
// Add SNAT mark if the packet is sent from local Pod to external addresses.
table=70, priority=0,reg0=0x2/0xffff,ip actions=load:0x1->NXM_NX_REG0[17], resubmit(,90)

Table105:
table=105, priority=200,ip,reg0=0x20000/0x20000, ct_state=+new+trk, actions=ct(commit,table=110,zone=65520,nat(src=${nodeIP}:10000-20000),exec(load:0x40->NXM_NX_CT_MARK[])))

Table110:
table=110, priority=200,ip,reg0=0x20000/0x20000 actions=output:gw0
```

> ### Use Windows named pipe for internal connections

Reviewer: s/Use/Using

> Named pipe is used for local connections on Windows Node instead of Unix Domain Socket(UDS). It is used in

Reviewer: s/on Windows Node/on Windows Node

> Named pipe is used for local connections on Windows Node instead of Unix Domain Socket(UDS). It is used in

Reviewer: space before (

these scenarios:
* OVSDB connection
* OpenFlow connection
* The connection between CNI plugin and CNI server
 
## Antrea Management on Windows

### Antrea Agent Management
Antrea Agent is running as a process on the Windows Node, but it is managed using a DaemonSet. The utility

> [Rancher Wins](https://github.com/rancher/wins) is used to manage the host process inside the container.

Reviewer: s/inside the container/from inside the DaemonSet Pod

> The configurations for Antrea Agent is configured using configmap, and the environment variables are set
> by kubelet on Windows.

Reviewer: ```suggestion
The Antrea Agent is configured using a ConfigMap, and the environment variables are set by kubelet on Windows.
```

### OVS Management

> OVS is working as Windows Service. Two corresponding services are registered: ovsdb-server and ovs-vswitchd.

Reviewer: OVS is running as 2 Windows Services: one for ovsdb-server and one for ovs-vswitchd.

> ## Traffic walk though

Reviewer: walkthrough

### Intra Cluster Traffic

> The traffic of Pod to local Pod and Pod to remote Pod is the same as Antrea on Linux. 

Reviewer: For consistency with the architecture document, use terms intra-Node Pod-to-Pod traffic and inter-Node Pod-to-Pod traffic

Table110:
table=110, priority=200,ip,reg0=0x20000/0x20000 actions=output:gw0
```

### Use Windows named pipe for internal connections
Named pipe is used for local connections on Windows Node instead of Unix Domain Socket (UDS). It is used in
these scenarios:
* OVSDB connection
* OpenFlow connection
* The connection between CNI plugin and CNI server
 
## Antrea Management on Windows

### Antrea Agent Management
Antrea Agent is running as a process on the Windows Node, but it is managed using a DaemonSet. The utility
[Rancher Wins](https://github.com/rancher/wins) is used to manage the host process inside the container.
The configurations for Antrea Agent is configured using configmap, and the environment variables are set
by kubelet on Windows.

### OVS Management
OVS is working as Windows Service. Two corresponding services are registered: ovsdb-server and ovs-vswitchd.

## Traffic walk though

### Intra Cluster Traffic

The traffic of Pod to local Pod and Pod to remote Pod is the same as Antrea on Linux. 
It is processed and forwarded by OVS, and controlled with OpenFlow entries.

### Service Traffic

> Kube-proxy userspace mode is configured to provide the Service functions. A specific Network Adapter named as

Reviewer: provide Service functions

> Kube-proxy userspace mode is configured to provide the Service functions. A specific Network Adapter named as

Reviewer: s/named as/named

"HNS Internal NIC" is provided to kube-proxy to configure Service addresses. The OpenFlow entries for the Service
traffic on Windows are the same as those on Linux.

The traffic of Pod to Service is similar to Antrea on Linux. The packet from Pod to Service leaves OVS pipeline
on the gateway interface first. kube-proxy captures the packet on the host and then chooses a valid endpoint. 

> The destination IP of the packet is changed to the endpoint, and enters the OVS pipeline again.

Reviewer: Handling of Pod-to-Service traffic is similar to Antrea on Linux. The packet from the Pod to the Service's ClusterIP leaves the OVS pipeline on the gateway interface first. kube-proxy captures the packet on the host and then chooses a valid endpoint. The destination IP of the packet is changed from the ClusterIP to the endpoint's IP, and the packet enters the OVS pipeline again.

A difference is the packet is using the client Pod's IP as source address when it enters the OVS pipeline at the
second time on Linux, but it uses the gateway IP as source on Windows. This is because kube-proxy on Windows sets
up a new connection on the gateway interface to the chosen backend Pod, and the packet is forwarded back to OVS

> from the new connection.

Reviewer: One difference is that on Linux, the packet is using the client Pod's IP as source address when it re-enters the OVS pipeline, while on Windows, it uses the gateway IP as source. This is because userspace kube-proxy on Windows terminates the initial connection and sets up a new connection on the gateway interface to the chosen backend Pod, after which the packet is forwarded back to OVS.

### External Traffic

> The traffic of Pod to external address is SNATed in OVS using the Node internal IP first, and then it leaves

Reviewer: Pod-to-external traffic

Node's internal IP

the OVS pipeline on the gateway interface. As IP-Forwarding on the gateway interface is enabled, the packet

> is forwarded to the OVS bridge interface and enters the OVS for the second time. 

Reviewer: s/the OVS/OVS

> When the packet enters the OVS at the second time, the dMAC is changed to an appropriate value using the host

Reviewer: s/the OVS/OVS

network stack. And OVS outputs the packet to the uplink interface directly.

For the reply packet of Pod to external address, it enters the OVS from the uplink interface. It will enter
the ct zone, and the destination IP address is recovered to the Pod IP within the ct context. Then it will be

> output to the Pod.

Reviewer: We can simplify:

The corresponding reply traffic will enter OVS on the uplink interface. It will access the ct zone, and the destination IP address will be translated back to the original Pod's IP within the ct context. Then it will be output to that Pod.

<img src="/docs/assets/windows_external_traffic.svg" width="600" alt="Traffic to external">

> ### Host Traffic

Reviewer: There may be a bit of redundancy with the "OVS uplink interface configuration" section above, but that's not a big deal

With “Transparent” mode, Antrea agent should also processes the host traffic if necessary, including the packets

> sent from the host to external, and the ones sent from the external to the host. 

Reviewer: In "Transparent" mode, the Antrea Agent should also process the host traffic when necessary, which includes packets sent from the host to external addresses, and the ones sent from external addresses to the host.

The host traffic is similar to the Pod external traffic, for the destination IP is not in the cluster CIDR, but
it is not sent from local Pod. OVS check the packet source firstly, if it is sent from the OVS bridge, and it is
not sent to local Pod, it is output to the uplink interface directly. The reply packet is checked in the ct
context first, then goes through the OVS pipeline if it has been set with ct_mark. The packet that is not marked

> in the ct context is output to the Windows host from OVS bridge interface. 

Reviewer: The host traffic case is similar to the Pod-to-external traffic, as the destination IP is not in the cluster CIDR, but we can differentiate based on the source IP. If the packet is sent from the OVS bridge interface, and if it is not destined to a local Pod, it is output to the uplink interface directly. Reply traffic goes through conntrack first, then through the rest of OVS pipeline if the ct_mark is set. If the mark is not set, it is output to the Windows host on the OVS bridge interface.

## Future Work

> 1. Support NetworkPolicy for Service traffic. Since kube-proxy in **userspace** mode will change the source

Reviewer: @wenyingd @ruicao93 can we open an issue on Github for this case to keep track of this?

Author: Done.

> IP in the request request packet, the NetworkPolicy rules doesn't work as expected. As Kubernetes Service is

Reviewer: extra "request" here

Author: Done.

> planned to be provided based on OpenFlow in Antrea, this scenario should be supported at that time.

Reviewer: s/should/will

Author: Done.

> planned to be provided based on OpenFlow in Antrea, this scenario should be supported at that time.

Reviewer: maybe add a link to the relevant issue for implementing ClusterIP in OVS

Author: Sure, will add the link for the issue.

# Running Antrea on Windows

> Antrea supports running on Windows worker Nodes. In this case, the Windows Node must

Reviewer 2: I think no need to repeat "the Windows Node must be a worker", and whether master is Linux is irrelevant to Antrea? Maybe remove the sentence of "In this case...".

Author: Sure, I will remove it.

assumes CNIAdd is an idempotent event, and it uses this event to query the Pod networking status.
Antrea needs to identify the container type (sandbox or workload) from the CNIAdd request:
* we create the HNS Endpoint only when the request is for the sandbox container
* we attach the HNS Endpoint no matter whether it is a sandbox container or a workload container.

### Gateway port configuration

The gateway port is created during the Antrea Agent initialization phase, and the address of the interface
should be the first IP in the subnet. The port is an OVS internal port and its default name is "gw0".

The gateway port is used to help implement L3 connectivity for the containers, including Pod-to-external,
Pod-to-Service, and Node-to-Pod. For Pod-to-external, and Pod-to-Service cases, OpenFlow entries are
installed in order to output these packets to the host on the gateway port. To ensure the packet is forwarded
correctly on the host, the IP-Forwarding feature should be enabled on the network adapter of the
gateway port.

A routing entry for traffic from the Node to the local Pod subnet is needed on the Windows host to ensure 

> that the packet can enter the OVS pipeline on the gateway port. This routing entry is added when the "gw0"

Reviewer: s/the "gw0"/"gw0"

is created.

Every time a new Node joins the cluster, a host routing entry on the gateway port is required, and the
remote subnet CIDR should be routed with the remote gateway address as the nexthop.

### Tunnel port configuration

Tunnel port configuration should be similar to Antrea on Linux:

> * tunnel port is created after OVS bridge is added; 

Reviewer: s/tunnel port is created after OVS bridge is added;/the tunnel port is added after the OVS bridge is created;

* a flow-based tunnel with the appropriate remote address is created for each Node in the cluster with OpenFlow.

The only difference with Antrea on Linux is that the tunnel local address is required when creating the tunnel
port (provided wth `local_ip` option). This local address is the one configured on the OVS bridge. 

### OVS bridge interface configuration

Since OVS is also responsible for taking charge of the network of the host, an interface for the OVS bridge
is required on which the host network settings are configured. It is created and enabled when creating
the OVS bridge, and the MAC address should be changed to be the same as the uplink interface. Then the IP 
address and the route entries originally configured on the uplink interface should also be migrated to
the interface.

> The packet that is sent to/from the Windows host should be forwarded on this interface. So the OVS bridge

Reviewer: s/The packet that is sent to/from the Windows host/The packets that are sent to/from the Windows host

is also a valid entry point into the OVS pipeline. A special ofport number 65534 (named as LOCAL) for the
OVS bridge is used in OpenFlow spec. 

In the OVS `Classifier` table, new OpenFlow entries are needed to match the packets from this interface.
There are two kinds of packets entering OVS pipeline from this interface:
 1) the packet is sent from the Windows host to an external address
 2) the packet is forwarded from the Windows host to the local Pod (e.g., this happens if the traffic is initiated
  from a Pod to a NodePort Service, and the chosen Node is where the Pod is located)

> For 1), the packet is always output to the uplink interface directly. For 2), the packet is injected into

Reviewer: bad markdown rendering because of missing empty line

the OVS pipeline and output to the backend Pod finally.
 
### OVS uplink interface configuration

After the OVS bridge is created, the original physical adapter is added to the OVS bridge as the uplink interface.
The uplink interface is used to support traffic from Pods accessing external addresses. The packet is always

> output to the uplink interface if it is sent to and external address and is entering OVS from the bridge

Reviewer: s/sent to and external address/sent to an external address

interface. 

We should differentiate the traffic if it is entering OVS from the uplink interface in OVS `Classifier`

> table. There are two kinds of packets entering the OVS from the uplink interface: 

Reviewer: s/entering the OVS/entering the OVS bridge

 1) reply traffic for the traffic that is sent from local Pods to external addresses,
 2) traffic on the host network

> For 1), the packets should enter the OVS pipeline and be written using SNAT information in connection tracking

Reviewer: same as above: bad markdown rendering because of missing empty line

> For 1), the packets should enter the OVS pipeline and be written using SNAT information in connection tracking

Reviewer: s/written/re-written

context first, and then processed by the OVS pipeline.
For 2), the packets are output to the OVS bridge interface directly. 

### SNAT configuration
SNAT is an important feature of the Antrea Agent on Windows Nodes, required to support Pods accessing external
addresses. It is implemented using OpenFlow. 

To support this feature, two additional marks are introduced:
* The 17th bit of NXM Register0 is set  for Pod-to-external traffic. This bit is set in the `L3Forwarding` table,
 and is consumed in the `ConntrackCommit` table.
* The SNATed traffic is marked with **0x40** in the ct context. This mark is set in the `ConntrackCommit`
 table when committing the new connection into the ct zone, and is consumed in the `ConntrackState` table
 to identify SNAT relevant packets.

The following changes in the OVS pipeline are needed:

> * Pick out the packet that is sent from local Pod-to-external address and set the SNAT bit in the register

Reviewer: s/Pick out/Indentify

 in the `L3Forwarding` table. The packet should have these characteristics: 

>   1) it enters the OVS pipeline from local Pod,

Reviewer: from a local Pod

  2) the destination IP address does not meet any of these conditions: cluster IP (local or any of the remote
  Subnet CIDR), or Node internal IP,
  3) the traffic is not for Service.

> * Implement SNAT rules and commit the new connection to the ct context in the `ConntrackCommit`  table. The SNAT

Reviewer: extra whitespace between `ConntrackCommit` and table

table=70,priority=190,ip,nw_dst=10.10.0.0/24,actions=resubmit(,80)
// Add SNAT mark if it is Pod-to-external traffic.
table=70, priority=0,reg0=0x2/0xffff,ip actions=load:0x1->NXM_NX_REG0[17], resubmit(,90)

ConntrackCommit Table: 105
table=105, priority=200,ip,reg0=0x20000/0x20000, ct_state=+new+trk, actions=ct(commit,table=110,zone=65520,nat(src=${nodeIP}:10000-20000),exec(load:0x40->NXM_NX_CT_MARK[])))

L2ForwardingOut Table: 110
table=110, priority=200,ip,reg0=0x20000/0x20000 actions=output:gw0
```

### Using Windows named pipe for internal connections
Named pipe is used for local connections on Windows Nodes instead of Unix Domain Socket (UDS). It is used in
these scenarios:
* OVSDB connection
* OpenFlow connection
* The connection between CNI plugin and CNI server
 
## Antrea Management on Windows

### Antrea Agent Management
The Antrea Agent is running as a process on the Windows Node, but it is managed using a DaemonSet. The utility
[Rancher Wins](https://github.com/rancher/wins) is used to manage the host process from inside the DaemonSet Pod.
The Antrea Agent is configured using a ConfigMap, and the environment variables are set by kubelet on Windows.

### OVS Management

> OVS is running as 2 Windows Service: one for ovsdb-server and one for ovs-vswitchd.

Reviewer: s/2 Windows Service/2 Windows Services

## Traffic walkthough

### Pod-to-Pod Traffic

The intra-Node Pod-to-Pod traffic and inter-Node Pod-to-Pod traffic are the same as Antrea on Linux. 
It is processed and forwarded by OVS, and controlled with OpenFlow entries.

> 

Reviewer: extra empty line on purpose?

### Service Traffic

Kube-proxy userspace mode is configured to provide Service functions. A specific Network Adapter named
"HNS Internal NIC" is provided to kube-proxy to configure Service addresses. The OpenFlow entries for the Service
traffic on Windows are the same as those on Linux.

> The Pod-to-Service traffic is similar to Antrea on Linux. The packet from Pod to Service leaves OVS pipeline

Reviewer: ```suggestion
The Pod-to-Service case is similar to Antrea on Linux. A packet from a Pod destined to a Service leaves the OVS pipeline
```

on the gateway interface first. kube-proxy captures the packet on the host and then chooses a valid endpoint. 
The destination IP of the packet is changed to the endpoint, and enters the OVS pipeline again.

One difference is that on Linux, the packet is using the client Pod's IP as source address when it re-enters the
OVS pipeline, while on Windows, it uses the gateway IP as source. This is because userspace kube-proxy on Windows
terminates the initial connection and sets up a new connection on the gateway interface to the chosen backend Pod, 

> after which the packet is forwarded back to OVS..

Reviewer: `.` or `...`?

### External Traffic

The Pod-to-external traffic is SNATed in OVS using the Node's internal IP first, and then it leaves
the OVS pipeline on the gateway interface. As IP-Forwarding on the gateway interface is enabled, the packet
is forwarded to the OVS bridge interface and enters OVS for the second time. 
When the packet enters OVS at the second time, the dMAC is changed to an appropriate value using the host
network stack. And OVS outputs the packet to the uplink interface directly.

> The corresponding reply traffic will enter OVS on the uplink interface. It will will access the ct zone, and

Reviewer: there is an extra "will" here

the destination IP will be translated back to the original Pod's IP within the ct context. Then it will be
output to the Pod.

<img src="/docs/assets/windows_external_traffic.svg" width="600" alt="Traffic to external">

### Host Traffic

> In “Transparent” mode, the Antrea Agent should also process the host traffic when necessary, which includes

Reviewer: fix double quotes

> packets sent from the host to external addresses, and the ones sent from the external addresses to the host. 

Reviewer: ```suggestion
packets sent from the host to external addresses, and the ones sent from external addresses to the host. 
```

> The host traffic case is similar to the Pod-to-external traffic, as the destination IP is not in the cluster

Reviewer: s/similar to the Pod-to-external traffic/similar to Pod-to-external traffic

CIDR, but we can differentiate based on the source IP. If the packet is sent from the OVS bridge interface,
and if it is destined to a local Pod, it is output to the uplink interface directly. Reply traffic goes through
conntrack first, then through the rest of the OVS pipeline if the ct_mark is set. If the packet is not set, it
is output to the Windows host on the OVS bridge interface. 

## Future Work
1. Support NetworkPolicy for Service traffic. Since kube-proxy in **userspace** mode will change the source

> IP in the request packet, the NetworkPolicy rules don't work as expected. As Kubernetes Service is

Reviewer 2: Nit: IP of the request packet.

Author: Done.

> planned to be implemented based on OpenFlow in Antrea ([Issue 463](https://github.com/vmware-tanzu/antrea/issues/463)),

Reviewer 2: Consider rephrasing this sentence. Like:
This issue will be fixed when Antrea implements [Kubernetes Service with OVS](...).

Author: Done.
